### DISCUSSION TITLE
mini.pick file picker with sorting from mini.visits

### DISCUSSION THREAD

**Discussion by JulesNP** [nvim-mini/mini.nvim#609] [General]
I wanted to write a frecency file picker which shows all files which would normally be seen in `:Pick files`, but sorted by frecency using data from `mini.visits`.

I'm currently using the implementation below, but it currently doesn't work with files with uppercase letters in their name, as all the paths passed in `stritems` to `match()` seem to be lowercase. As a result, they don't match the table keys from `MiniVisits.get_index()`. 

What would be a good workaround for this?

```lua
MiniPick.registry.frecency = function()
    local visits_index = MiniVisits.get_index()[vim.loop.cwd()]
    if visits_index == nil then
        MiniPick.builtin.files()
    else
        local current_file = vim.fn.expand "%:p"
        MiniPick.builtin.files(nil, {
            source = {
                match = function(stritems, indices, query)
                    local prompt = vim.pesc(table.concat(query))
                    local res = {}

                    for _, index in ipairs(indices) do
                        local path = stritems[index]
                        local match_score = prompt == "" and 0 or MiniFuzzy.match(prompt, path).score
                        if match_score >= 0 then
                            local full_path = vim.fn.fnamemodify(path, ":p")
                            -- NOTE: Currently doesn't match files with uppercase characters
                            local visits = visits_index[full_path]
                            table.insert(res, {
                                count = visits == nil and 0 or visits.count,
                                index = index,
                                latest = visits == nil and 0 or visits.latest,
                                score = full_path == current_file and 9999 or match_score, -- Score current file last
                            })
                        end
                    end

                    local function set_rank(key)
                        local last, ties = -1, 0
                        for index, item in ipairs(res) do
                            local value = item[key]
                            ties = last == value and ties + 1 or 0
                            last = value
                            item[key] = index - ties - 1
                        end
                    end

                    table.sort(res, function(a, b)
                        return a.count > b.count
                    end)
                    set_rank "count"

                    table.sort(res, function(a, b)
                        return a.latest > b.latest
                    end)
                    set_rank "latest"

                    table.sort(res, function(a, b)
                        return (a.score + a.count + a.latest) < (b.score + b.count + b.latest)
                    end)

                    return vim.tbl_map(function(val)
                        return val.index
                    end, res)
                end,
            },
        })
    end
end
```

**Comment by echasnovski** [comment]
That is certainly an interesting idea to implement. Nice job trying to implement it!

Sorry, it is a bit convoluted to review in full, but here are the steps I'd take to implement this:

- Use [`MiniVisits.list_paths()`](https://github.com/echasnovski/mini.nvim/blob/c87d3b73d763d4222cdef19323e6a2c4bee33013/doc/mini-visits.txt#L540) directly to get an array of paths already sorted in the order you'd like.
- Invert it so that paths are keys and their id is a value.
- Update `match()` to first check that table and possibly sort based on it.

> I'm currently using the implementation below, but it currently doesn't work with files with uppercase letters in their name, as all the paths passed in `stritems` to `match()` seem to be lowercase. As a result, they don't match the table keys from `MiniVisits.get_index()`.

That [depends on the query and 'ignorecase'/'smartcase' options](https://github.com/echasnovski/mini.nvim/blob/c87d3b73d763d4222cdef19323e6a2c4bee33013/doc/mini-pick.txt#L326-L331). If you need to compare to the data from 'mini.visits', then possibly convert *that* to lowercase (if 'mini.pick' uses it in that particular call).

**Comment by JulesNP** [comment]
Thanks! That makes sense, and would cut out a lot of the extra scoring logic in my implementation.

**Reply by Susensio** [reply]
i you get a working implementation, please, share! I would love a frecency sorting as well

**Reply by JulesNP** [reply]
Sure! This is what I'm using currently:

```lua
MiniPick.registry.frecency = function()
    local visit_paths = MiniVisits.list_paths()
    local current_file = vim.fn.expand "%"
    MiniPick.builtin.files(nil, {
        source = {
            match = function(stritems, indices, query)
                -- Concatenate prompt to a single string
                local prompt = vim.pesc(table.concat(query))

                -- If ignorecase is on and there are no uppercase letters in prompt,
                -- convert paths to lowercase for matching purposes
                local convert_path = function(str)
                    return str
                end
                if vim.o.ignorecase and string.find(prompt, "%u") == nil then
                    convert_path = function(str)
                        return string.lower(str)
                    end
                end

                local current_file_cased = convert_path(current_file)
                local paths_length = #visit_paths

                -- Flip visit_paths so that paths are lookup keys for the index values
                local flipped_visits = {}
                for index, path in ipairs(visit_paths) do
                    local key = vim.fn.fnamemodify(path, ":.")
                    flipped_visits[convert_path(key)] = index - 1
                end

                local result = {}
                for _, index in ipairs(indices) do
                    local path = stritems[index]
                    local match_score = prompt == "" and 0 or MiniFuzzy.match(prompt, path).score
                    if match_score >= 0 then
                        local visit_score = flipped_visits[path] or paths_length
                        table.insert(result, {
                            index = index,
                            -- Give current file high value so it's ranked last
                            score = path == current_file_cased and 999999 or match_score + visit_score * 10,
                        })
                    end
                end

                table.sort(result, function(a, b)
                    return a.score < b.score
                end)

                return vim.tbl_map(function(val)
                    return val.index
                end, result)
            end,
        },
    })
end
```

**Reply by bassamsdata** [reply]
Thank you for this! Been using it for a week, works like a charm with no issues. Speed is excellent. Replaced the original `:pick files` with this one; sorting in the original `files` means nothing. I'd Suggest including it in the plugin itself tbh, so other people can benefit from it.

**Reply by Susensio** [reply]
I'm having problems when query contains spaces. For now, I workaround by removing spaces with
```lua
              prompt = string.gsub(prompt, "%s+", "")

```

**Reply by JulesNP** [reply]
@Susensio, can you advise what issue you're having with spaces in the query? For me entering a space filters the list to only files with spaces in the name. (Which is usually none for most of my codebases)

<img width="488" alt="image" src="https://github.com/echasnovski/mini.nvim/assets/44929900/7bee86f6-2197-4b26-b64e-b4bd25c77590">

**Reply by Susensio** [reply]
Hey @JulesNP , sorry for the late reply.
My problem is, as stated previously, that special characters behave differently in MiniFuzzy. With the default file picker, spaces do not match directly, and I cannot get dots to match with MiniFuzzy.
Maybe @echasnovski can give us some indications on the differences between MiniPick matcher and MiniFuzzy matcher. How can we make them behave consistently?

**Reply by echasnovski** [reply]
> Maybe @echasnovski can give us some indications on the differences between MiniPick matcher and MiniFuzzy matcher. How can we make them behave consistently?

There is no easy way, unfortunately (if at all). I initially planned for matching in 'mini.pick' to be the same as in 'mini.fuzzy'. In the end, I got a bit carried away when saw that implementing extra rules (like `^`, `$`, `'`, and special treatment of spaces) has more benefits than added complexity.

I may look at having 'mini.fuzzy' behave the same as in 'mini.pick' in the future. Doubt that it will happen soon, though.

**Reply by Susensio** [reply]
I'm trying to understand the logic behind your matching algorithm. Can I ask you a couple of questions?
1 - How does MiniPick handles spaces?
2 - Why can't I match a dot `.` with MiniFuzzy?

**Reply by echasnovski** [reply]
> 1 - How does MiniPick handles spaces?

Basically, item's string should match each non-whitespace part exactly and in that particular order. So query `ab c` will match `_ab__c` but not `_a_b_c`.

Here us a [full description of default matching](https://github.com/echasnovski/mini.nvim/blob/a683bfe8e03293e3bb079e24c94e51977756a3ec/doc/mini-pick.txt#L848) with [relevant example](https://github.com/echasnovski/mini.nvim/blob/a683bfe8e03293e3bb079e24c94e51977756a3ec/doc/mini-pick.txt#L907).. Here is also a [short one](https://github.com/echasnovski/mini.nvim/blob/a683bfe8e03293e3bb079e24c94e51977756a3ec/doc/mini-pick.txt#L184-L191) (does not mention spaces, though).

> 2 - Why can't I match a dot `.` with MiniFuzzy?

You can. Try `:=require('mini.fuzzy').filtersort('.', { 'abc', 'ab.c', 'a.bc' })`.

**Comment by Susensio** [comment]
> > 1 - How does MiniPick handles spaces?
> 
> Basically, item's string should match each non-whitespace part exactly and in that particular order. So query `ab c` will match `_ab__c` but not `_a_b_c`.
> 
> Here us a [full description of default matching](https://github.com/echasnovski/mini.nvim/blob/a683bfe8e03293e3bb079e24c94e51977756a3ec/doc/mini-pick.txt#L848) with [relevant example](https://github.com/echasnovski/mini.nvim/blob/a683bfe8e03293e3bb079e24c94e51977756a3ec/doc/mini-pick.txt#L907).. Here is also a [short one](https://github.com/echasnovski/mini.nvim/blob/a683bfe8e03293e3bb079e24c94e51977756a3ec/doc/mini-pick.txt#L184-L191) (does not mention spaces, though).

I see. Is there any way of calling default_match without sorting?


> > 2 - Why can't I match a dot `.` with MiniFuzzy?
> 
> You can. Try `:=require('mini.fuzzy').filtersort('.', { 'abc', 'ab.c', 'a.bc' })`.

Okey, the problem was that `vim.pesc` was being executed twice, once in my code and other in MiniFuzzy.


------------------

Any way, I've done a rewrite of this. To improve consistency, I've managed to use `MiniPick.default_match` for filtering, and `MiniVisits.list_paths` for sorting.
That is, always sort by frecency. This drops the `MiniFuzzy` dependency.
@JulesNP Maybe you find this valuable. It does not have any problem with spaces nor dots (as far as I've tested).

```lua
      MiniPick.registry.frecency = function()
        local inf = math.huge

        local visit_paths = MiniVisits.list_paths()
        visit_paths = vim.tbl_map(function(path) return vim.fn.fnamemodify(path, ":.") end, visit_paths)
        vim.tbl_add_reverse_lookup(visit_paths)

        -- current file last
        local current_file = vim.fn.expand("%:.")
        if visit_paths[current_file] then
          visit_paths[current_file] = inf
        end

        MiniPick.builtin.files(nil, {
          source = {
            name = "Files (MRU)",
            match = function(stritems, indices, query)
              local filtered = MiniPick.default_match(stritems, indices, query, true) or {}

              table.sort(filtered,
                function(item1, item2)
                  local path1 = stritems[item1]
                  local path2 = stritems[item2]
                  local score1 = visit_paths[path1] or inf    -- inf for paths not visited
                  local score2 = visit_paths[path2] or inf    -- inf for paths not visited
                  return score1 < score2
                end
              )

              return filtered
            end,
          },
        })
      end
```

I don't know how could this behave with a huge amount of files, as I'm calling `default_match` synchronously. Maybe @echasnovski can give us some directions on a better integration.

**Reply by echasnovski** [reply]
> I see. Is there any way of calling default_match without sorting?

Not at the moment, no. The best alternative is to do a synchronous call and resort later.

> I don't know how could this behave with a huge amount of files, as I'm calling `default_match` synchronously. Maybe @echasnovski can give us some directions on a better integration.

In my experience, on modern hardware anything under 10000 items should be fine. So I'd stick with doing synchronous match with later sorting.

**Reply by Susensio** [reply]
Thanks for your help!
Would you be interested in making this a Mini.extra picker? I could try to do a proper PR.

**Reply by echasnovski** [reply]
Thanks for the offer, but I think having `MiniExtra.pickers.visits` is enough.

**Reply by Susensio** [reply]
Got it. The problem I had with `MiniExtra.pickers.visits` is that it does not show unvisited files, so it cannot be the one and only file picker.

**Reply by wiseodd** [reply]
Thanks for sharing @Susensio! I can report that it works great. I would replace the default sorting algorithm with the `z`-like sorting that is also available in the plugin:

```lua
...
local sort = MiniVisits.gen_sort.z()
local visit_paths = MiniVisits.list_paths(nil, { sort = sort })
...
```

I find it behaves more similarly to [telescope smart-open](https://github.com/danielfalk/smart-open.nvim), which is, dare I say, the gold standard for this "omnibar"-style picker.

**Reply by Susensio** [reply]
I'm trying to implement a match_filter function just like [zoxide](https://github.com/ajeetdsouza/zoxide/wiki/Algorithm), that is, that prioritizes last path component:

But I'm bumping into what I think is some kind of cache of `mini.pick`, maybe @echasnovski can give me a hint.

https://asciinema.org/a/9J7cHXeMoyyJNG6tTHwOgX959
The problem is, when looking for `lang python`, results wont get updated until I backspace. I'm guessing that `mini.pick` tries to keep a cache and invalidates it on backspace, but I cannot manage to disable that behaviour. I've tried `use_cache = false` with no luck.

This is my full code:

```lua
         MiniPick.registry.frecency = function()
            local function sort(stritems, indices)
               local inf = math.huge

               local minivisits = require("mini.visits")
               local visit_paths = minivisits.list_paths(nil, { sort = minivisits.gen_sort.z() })

               -- relative paths to cwd
               visit_paths = vim.tbl_map(function(path) return vim.fn.fnamemodify(path, ":.") end, visit_paths)
               local scores = {}
               for i, path in ipairs(visit_paths) do
                  scores[path] = i
               end

               -- current file last
               local current_file = vim.fn.expand("%:.")
               if scores[current_file] then
                  scores[current_file] = inf
               end

               indices = vim.deepcopy(indices)
               table.sort(indices,
                  function(item1, item2)
                     local path1 = stritems[item1]
                     local path2 = stritems[item2]

                     local score1 = scores[path1] or inf -- inf for paths not visited
                     local score2 = scores[path2] or inf -- inf for paths not visited

                     return score1 < score2
                  end
               )
               return indices
            end

            local function get_keyword_matches(path, keywords)
               if #keywords == 0 then
                  return {} -- No keywords, return empty list
               end

               path = path:lower() -- Convert to lowercase for case-insensitive matching
               local matches = {}

               -- Get the last keyword and remaining ones
               local keywords_last = keywords[#keywords]
               local keywords_rest = vim.list_slice(keywords, 1, #keywords - 1)

               local last_path_separator = path:match("^.*()" .. "/") or 0
               local last_start, last_end = path:find(keywords_last, last_path_separator + 1, true)

               if last_start == nil then
                  return nil -- Last keyword not found
               end

               -- Store last keyword match
               table.insert(matches, { last_start, last_end })

               path = path:sub(1, last_start - 1) -- Truncate before last keyword

               -- Find remaining keywords in reverse order
               for i = #keywords_rest, 1, -1 do
                  local keyword = keywords_rest[i]
                  local start_idx, end_idx = path:find(keyword, 1, true)
                  if not start_idx then
                     return nil -- Missing keyword
                  end

                  -- Store match
                  table.insert(matches, { start_idx, end_idx })

                  -- Truncate path before this keyword
                  path = path:sub(1, start_idx - 1)
               end

               -- Reverse order of matches (since we inserted in reverse)
               return matches
            end

            local function match_filter(stritems, indices, query)
               if #query == 0 or query == nil then
                  return indices
               end
               local keywords = vim.split(table.concat(query), " ")
               if #keywords == 0 or keywords == nil then
                  return indices
               end
               local filtered = {}
               for _, index in ipairs(indices) do
                  local item = stritems[index]
                  if item ~= nil then
                     local match = get_keyword_matches(item, keywords)
                     if match ~= nil then
                        table.insert(filtered, index)
                     end
                  end
               end
               return filtered
            end

            local function show(buf_id, items, query, opts)
               local ns = vim.api.nvim_get_namespaces()["MiniPickRanges"]

               local icons = {}
               local lines = {}
               for i, item in ipairs(items) do
                  local icon_text, icon_hl_group = require("mini.icons").get("file", item)
                  icons[i] = { text = icon_text, hl_group = icon_hl_group }
                  lines[i] = icons[i].text .. ' ' .. item
               end

               vim.api.nvim_buf_set_lines(buf_id, 0, -1, false, lines)

               -- vim.api.nvim_buf_clear_namespace(buf_id, ns, 0, -1)

               -- Highlight icons
               for i, icon in ipairs(icons) do
                  local row = i - 1
                  vim.api.nvim_buf_set_extmark(buf_id, ns, row, 0, {
                     end_col = 1,
                     end_row = row,
                     hl_group = icon.hl_group,
                     hl_mode = "combine",
                     priority = 200,
                  })
               end

               -- Highlight matched ranges
               for i, item in ipairs(items) do
                  local keywords = vim.split(table.concat(query), " ")
                  local ranges = get_keyword_matches(item, keywords)
                  if ranges == nil then
                     return
                  end
                  for _, range in ipairs(ranges) do
                     local row = i - 1
                     local start_col = range[1] - 1 + icons[i].text:len() + 1
                     local end_col = range[2] + icons[i].text:len() + 1
                     vim.api.nvim_buf_set_extmark(buf_id, ns, row, start_col, {
                        end_col = end_col,
                        end_row = row,
                        hl_group = "MiniPickMatchRanges",
                        hl_mode = "combine",
                        priority = 200,
                     })
                  end
               end
            end

            MiniPick.builtin.files(nil, {
               source = {
                  name = "Files (MRU)",
                  match = function(stritems, indices, query)
                     indices = match_filter(stritems, indices, query)
                     indices = sort(stritems, indices)
                     return indices
                  end,
                  show = show,
               },
               options = {
                  use_cache = false,
               },
            })
         end

```

**Reply by echasnovski** [reply]
> But I'm bumping into what I think is some kind of cache of `mini.pick`, maybe @echasnovski can give me a hint.
> 
> https://asciinema.org/a/9J7cHXeMoyyJNG6tTHwOgX959 The problem is, when looking for `lang python`, results wont get updated until I backspace. I'm guessing that `mini.pick` tries to keep a cache and invalidates it on backspace, but I cannot manage to disable that behaviour. I've tried `use_cache = false` with no luck.

I think it is a result of [this behavior](https://github.com/echasnovski/mini.nvim/blob/4dd209e4f03bacd46d6d87408b2529a8823112fb/doc/mini-pick.txt#L322-L326). The `inds` in `source.match()` *usually* point to the the *latest* matches (i.e. matches before the current picker "update cycle") for better performance. Pressing `<BS>` resets them and `source.match()` will be called for all `stritems`.

In the demo, there is no '.../lang/python.lua' file in the matches as early as at the "la" query. As every next key press will only narrow these matches, there will be no '.../lang/python.lua' at "lang python" query. Pressing `<BS>` makes `source.match()` run on every `stritems` which (for some reason) now matches '.../lang/pythong.lua' file.

So either try and find why that file didn't match at "la" query (which I'd think it should have) or ignore second argument to custom `source.match()` and always perform matching on all `stritems`.

**Reply by timofurrer** [reply]
This is super useful! Is there any change that we can get such a "mru" / frequency-based sorted picker into mini pick directly? 🙏

**Reply by echasnovski** [reply]
> Is there any change that we can get such a "mru" / frequency-based sorted picker into mini pick directly? 🙏

For files in 'mini.pick' directly - probably no, since it is not *that* trivial to do.

For files in 'mini.extra' - *might be*. It already contains [`MiniExtra.pickers.visit_paths()`](https://nvim-mini.org/mini.nvim/doc/mini-extra.html#miniextra.pickers.visit_paths) with ordering from 'mini.visits' which is usually what is needed anyways (in my experience).

And in general I feel that sorting in pickers other than "whichever matches closer to the query" is rarely an improvement. The only case thus far is when there is an inherent ordering in items, like buffer lines or already sorted `visit_paths`. Those contain `preserve_order` optional flag.
